# 2 Data Wrangling<a id='2_Data_Wrangling'></a>

## 2.1 Contents<a id='2.1_Contents'></a>
* [2 Data wrangling](#2_Data_wrangling)
  * [2.1 Contents](#2.1_Contents)
  * [2.2 Introduction](#2.2_Introduction)
  * [2.3 Imports](#2.3_Imports)
  * [2.4 Descriptive Statistics](#2.4_Descriptive_Stat)

## 2.2 Introduction<a id='2.2_Introduction'></a>

This steps focuses on collecting data, organizing it, and making sure it's well defined before cleaning and explore more your data.

## 2.3 Imports<a id='2.3_Import'></a>

_*Lets start.. here we are importing the dataset.*_

In [15]:
#Import pandas, matplotlib.pyplot, and seaborn in the correct lines below
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [2]:
#Get the current working directory of a process and print 
MyWorkingDir = os.getcwd()
print(MyWorkingDir)

/Users/joeaquino/Downloads


In [3]:
# load and read the Dataset Cleveland.csv on woking directory
# and fed into DataFrame or df
path = MyWorkingDir = '/Users/joeaquino/Downloads/Cleveland.csv'
df = pd.read_csv('/Users/joeaquino/Downloads/Cleveland.csv', header = None)

## 2.4 Descriptive Statistics<a id='2.4_Descriptive_Stat'></a>

_*In this part we are generating summaries of our Cleveland.csv dataset*_

In [4]:
#Called the info method on df to see a summary of the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       303 non-null    int64  
 1   1       303 non-null    int64  
 2   2       303 non-null    int64  
 3   3       303 non-null    int64  
 4   4       303 non-null    int64  
 5   5       303 non-null    int64  
 6   6       303 non-null    int64  
 7   7       303 non-null    int64  
 8   8       303 non-null    int64  
 9   9       303 non-null    float64
 10  10      303 non-null    int64  
 11  11      303 non-null    object 
 12  12      303 non-null    object 
 13  13      303 non-null    int64  
dtypes: float64(1), int64(11), object(2)
memory usage: 33.3+ KB


In [5]:
#Call the describe method on df to get more statictical measurements.
df.describe()

0           1           2           3           4           5   \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.438944    0.679868    3.158416  131.689769  246.693069    0.148515   
std      9.038662    0.467299    0.960126   17.599748   51.776918    0.356198   
min     29.000000    0.000000    1.000000   94.000000  126.000000    0.000000   
25%     48.000000    0.000000    3.000000  120.000000  211.000000    0.000000   
50%     56.000000    1.000000    3.000000  130.000000  241.000000    0.000000   
75%     61.000000    1.000000    4.000000  140.000000  275.000000    0.000000   
max     77.000000    1.000000    4.000000  200.000000  564.000000    1.000000   

               6           7           8           9           10          13  
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000  
mean     0.990099  149.607261    0.326733    1.039604    1.600660    0.937294  
std      0.994971   22.875003    0.469794    1.161075    0.616226    1.228536  
min      0.000000   71.000000    0.000000    0.000000    1.000000    0.000000  
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000  
50%      1.000000  153.000000    0.000000    0.800000    2.000000    0.000000  
75%      2.000000  166.000000    1.000000    1.600000    2.000000    2.000000  
max      2.000000  202.000000    1.000000    6.200000    3.000000    4.000000

In [6]:
#Call the head method on df to print the first 10 rows of the data.
df.head(10)

0   1   2    3    4   5   6    7   8    9   10 11 12  13
0  63   1   1  145  233   1   2  150   0  2.3   3  0  6   0
1  67   1   4  160  286   0   2  108   1  1.5   2  3  3   2
2  67   1   4  120  229   0   2  129   1  2.6   2  2  7   1
3  37   1   3  130  250   0   0  187   0  3.5   3  0  3   0
4  41   0   2  130  204   0   2  172   0  1.4   1  0  3   0
5  56   1   2  120  236   0   0  178   0  0.8   1  0  3   0
6  62   0   4  140  268   0   2  160   0  3.6   3  2  3   3
7  57   0   4  120  354   0   0  163   1  0.6   1  0  3   0
8  63   1   4  130  254   0   2  147   0  1.4   2  1  7   2
9  53   1   4  140  203   1   2  155   1  3.1   3  0  7   1

_*on first 10 rows it seems no missing values... Lets check the tail*_

In [7]:
#Call the tail method on df to print the last 10 rows of the data.
df.tail(10)

0   1   2    3    4   5   6    7   8    9   10 11 12  13
293  63   1   4  140  187   0   2  144   1  4.0   1  2  7   2
294  63   0   4  124  197   0   0  136   1  0.0   2  0  3   1
295  41   1   2  120  157   0   0  182   0  0.0   1  0  3   0
296  59   1   4  164  176   1   2   90   0  1.0   2  2  6   3
297  57   0   4  140  241   0   0  123   1  0.2   2  0  7   1
298  45   1   1  110  264   0   0  132   0  1.2   2  0  7   1
299  68   1   4  144  193   1   0  141   0  3.4   2  2  7   2
300  57   1   4  130  131   0   0  115   1  1.2   2  1  7   3
301  57   0   2  130  236   0   2  174   0  0.0   2  1  3   1
302  38   1   3  138  175   0   0  173   0  0.0   1  ?  3   0

_* check row 302 theres a "?" value. Let's try sample method*_

In [8]:
#Call the sample method on df to print the first ramdom 20 rows of the data.
df.sample(20)

0   1   2    3    4   5   6    7   8    9   10 11 12  13
255  42   0   3  120  209   0   0  173   0  0.0   2  0  3   0
94   63   0   3  135  252   0   2  172   0  0.0   1  0  3   0
257  76   0   3  140  197   0   1  116   0  1.1   2  0  3   0
265  42   1   4  136  315   0   0  125   1  1.8   2  0  6   2
212  41   1   3  130  214   0   2  168   0  2.0   2  0  3   0
251  58   1   4  146  218   0   0  105   0  2.0   2  1  7   1
134  43   0   3  122  213   0   0  165   0  0.2   2  0  3   0
283  35   1   2  122  192   0   0  174   0  0.0   1  0  3   0
214  52   1   4  112  230   0   0  160   0  0.0   1  1  3   1
178  43   1   3  130  315   0   0  162   0  1.9   1  1  3   0
285  58   1   4  114  318   0   1  140   0  4.4   3  3  6   4
172  59   0   4  174  249   0   0  143   1  0.0   2  0  3   1
230  52   0   3  136  196   0   2  169   0  0.1   2  0  3   0
296  59   1   4  164  176   1   2   90   0  1.0   2  2  6   3
276  66   0   3  146  278   0   2  152   0  0.0   2  1  3   0
3    37   1   3  130  250   0   0  187   0  3.5   3  0  3   0
6    62   0   4  140  268   0   2  160   0  3.6   3  2  3   3
25   50   0   3  120  219   0   0  158   0  1.6   2  0  3   0
116  58   1   3  140  211   1   2  165   0  0.0   1  0  3   0
29   40   1   4  110  167   0   2  114   1  2.0   2  0  7   3

*_"ca" has a value of "?" in columns 166, 192, 287 and 302_*

_* "thal" has a value of "?" in columns 87 and 266.*_

_* Lets change the rows values to given attributes as follows.*_

 Attribute Information:
 
      0. (age)       age in years
      1. (sex)       sex (1 = male; 0 = female)
      2. (cp)        chest pain type
                            -- Value 1: typical angina
                            -- Value 2: atypical angina
                            -- Value 3: non-anginal pain
                            -- Value 4: asymptomatic
      3. (trestbps)  resting blood pressure (in mm Hg on admission to the hospital)
      4. (chol)      serum cholestoral in mg/dl
      5. (fbs)       (fasting blood sugar > 120 mg/dl)  (1 = true; 0 = false)
      6. (restecg)   resting electrocardiographic results
                            -- Value 0: normal
                            -- Value 1: having ST-T wave abnormality (T wave inversions and/or ST
                                        elevation or depression of > 0.05 mV)
                            -- Value 2: showing probable or definite left ventricular hypertrophy
                                        by Estes' criteria
      7. (thalach)   maximum heart rate achieved
      8. (exang)     exercise induced angina (1 = yes; 0 = no)
      9.(oldpeak)   ST depression induced by exercise relative to rest
      10.(slope)     e 1: upsloping
                            -- Value 2: flat
                            -- Value 3: downsloping
      11. (ca)        number of major vessels (0-3) colored by flourosopy
      12. (thal)      3 = normal; 6 = fixed defect; 7 = reversable defect
      13. (num)       (the predicted attribute)
                            diagnosis of heart disease (angiographic disease status)
                                    -- Value 0: < 50% diameter narrowing
                                    -- Value 1: > 50% diameter narrowing
                                    (in any major vessel: attributes 59 through 68 are vessels)

In [9]:
# add headers to the data.
df.columns = ['age', 'sex', 'cp', 'trestbps', 'chol',
              'fbs', 'restecg', 'thalach', 'exang', 
              'oldpeak', 'slope', 'ca', 'thal', 'num']

In [10]:
# called the info method on df to see a summary of the data.
# and headers are now added
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    object 
 12  thal      303 non-null    object 
 13  num       303 non-null    int64  
dtypes: float64(1), int64(11), object(2)
memory usage: 33.3+ KB


In [16]:
#Checking Missing values.
df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
num         0
dtype: int64

Here we see that "ca" has 4 and thal has 2 possible missing values

In [17]:
# Calling count method to df to check possible missing values 
df.count()

age         303
sex         303
cp          303
trestbps    303
chol        303
fbs         303
restecg     303
thalach     303
exang       303
oldpeak     303
slope       303
ca          299
thal        301
num         303
dtype: int64

Here "ca" has 299 and thal has 301. We have 303 total entries and "ca" missing 4 and "thal" missing 2(see Cell 44)

In [18]:
#Replace '?' to NaN
df.replace("?",np.nan, inplace = True)
df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   1       145   233    1        2      150      0      2.3   
1     67    1   4       160   286    0        2      108      1      1.5   
2     67    1   4       120   229    0        2      129      1      2.6   
3     37    1   3       130   250    0        0      187      0      3.5   
4     41    0   2       130   204    0        2      172      0      1.4   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   45    1   1       110   264    0        0      132      0      1.2   
299   68    1   4       144   193    1        0      141      0      3.4   
300   57    1   4       130   131    0        0      115      1      1.2   
301   57    0   2       130   236    0        2      174      0      0.0   
302   38    1   3       138   175    0        0      173      0      0.0   

     slope   ca thal  num  
0        3    0    6    0  
1        2    3    3    2  
2        2    2    7    1  
3        3    0    3    0  
4        1    0    3    0  
..     ...  ...  ...  ...  
298      2    0    7    1  
299      2    2    7    2  
300      2    1    7    3  
301      2    1    3    1  
302      1  NaN    3    0  

[303 rows x 14 columns]

In [19]:
for column in df.columns.values.tolist():
    print(column)
    print(df[column].value_counts(dropna=False))

age
58    19
57    17
54    16
59    14
52    13
60    12
51    12
56    11
62    11
44    11
64    10
41    10
67     9
63     9
42     8
43     8
45     8
53     8
55     8
61     8
65     8
50     7
66     7
48     7
46     7
47     5
49     5
70     4
68     4
35     4
39     4
69     3
71     3
40     3
34     2
37     2
38     2
29     1
77     1
74     1
76     1
Name: age, dtype: int64
sex
1    206
0     97
Name: sex, dtype: int64
cp
4    144
3     86
2     50
1     23
Name: cp, dtype: int64
trestbps
120    37
130    36
140    32
110    19
150    17
138    12
128    12
160    11
125    11
112     9
132     8
118     7
124     6
108     6
135     6
152     5
134     5
145     5
100     4
170     4
122     4
126     3
136     3
115     3
180     3
142     3
105     3
102     2
146     2
144     2
148     2
178     2
94      2
165     1
123     1
114     1
154     1
156     1
106     1
155     1
172     1
200     1
101     1
129     1
192     1
158     1
104     1
174     1
117   

##### We will deal with "Nan" values on Exploratory Data Analysis